### PR TITLE
[WIP] mvcc:init command import for load data

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
    This PR implement command `CmdImportRequest` to support 1pc transaction  in loading data.

@disksing @zhangjinpeng1987 @hhkbp2 PTAL

### CODE REVIEW

File: tests/storage/test_storage.rs [hunk 1]
```diff
@@ -44,6 +44,13 @@ fn test_txn_store_get() {
 }
 
 #[test]
+fn test_import() {
+    let store = AssertionStorage::default();
+    store.import_ok(b"x", b"y", 5);
```
Comment (zhangjinpeng1987) [inline tests/storage/test_storage.rs]: s/x/k
s/y/v

**Comment by disksing** [comment]
Is it possible for _Import_ command to skip snapshot stage? As you don't read any values from Snapshot anyway.

**Comment by disksing** [comment]
Is it safe for _Import_ to skip the _Latch_? cc @zhangjinpeng1987

File: src/server/kv.rs [hunk 2]
```diff
@@ -199,6 +200,28 @@ impl StoreHandler {
...
+        if !msg.has_cmd_import_req() {
+            return Err(box_err!("msg doesn't contain a CmdImportRequest"));
+        }
+        let mut req = msg.take_cmd_import_req();
+        let mutations = req.take_mutations()
+            .into_iter()
+            .map(|mut x| {
+                match x.get_op() {
+                    Op::Put => Mutation::Put((Key::from_raw(x.get_key()), x.take_value())),
+                    Op::Del => Mutation::Delete(Key::from_raw(x.get_key())),
```
Comment (siddontang) [inline src/server/kv.rs]: Does import have del operation?

File: src/storage/txn/scheduler.rs [hunk 3]
```diff
@@ -570,6 +570,17 @@ fn process_write_impl(cid: u64,
                 (pr, txn.modifies())
             }
         }
+        Command::Import { ref mutations, commit_ts, .. } => {
+            let mut txn = MvccTxn::new(snapshot, &mut statistics, commit_ts, None);
+            for m in mutations {
+                match txn.import(m.clone(), commit_ts) {
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: Please use `try!` instead.

File: src/server/kv.rs [hunk 4]
```diff
@@ -199,6 +200,28 @@ impl StoreHandler {
...
+                }
+            })
+            .collect();
+        let cb = self.make_cb(StoreHandler::cmd_import_done, on_resp);
+        self.store
+            .async_import(msg.take_context(), mutations, req.get_commit_version(), cb)
+            .map_err(Error::Storage)
+    }
+
+
```
Comment (zhangjinpeng1987) [inline src/server/kv.rs]: Please remove blank line.

File: src/storage/mod.rs [hunk 5]
```diff
@@ -244,6 +256,7 @@ impl Command {
             Command::ScanLock { .. } => "scan_lock",
             Command::ResolveLock { .. } => "resolve_lock",
             Command::Gc { .. } => CMD_TAG_GC,
+            Command::Import { .. } => "import",
```
Comment (zhangjinpeng1987) [inline src/storage/mod.rs]: Please add `Command::Import` in function `readonly`

File: src/server/kv.rs [hunk 6]
```diff
@@ -208,10 +208,11 @@ impl StoreHandler {
...
             .map(|mut x| {
-                match x.get_op() {
-                    Op::Put => Mutation::Put((Key::from_raw(x.get_key()), x.take_value())),
-                    Op::Del => Mutation::Delete(Key::from_raw(x.get_key())),
-                    Op::Lock => Mutation::Lock(Key::from_raw(x.get_key())),
+                let op = x.get_op();
+                if op == Op::Put {
+                    Mutation::Put((Key::from_raw(x.get_key()), x.take_value()))
+                } else {
+                    panic!("CmdImportRequest does not support this Operation:{:?}", op);
```
Comment (siddontang) [inline src/server/kv.rs]: can't return error instead?

File: src/storage/mod.rs [hunk 7]
```diff
@@ -113,6 +113,11 @@ pub enum Command {
         lock_ts: u64,
         commit_ts: u64,
     },
+    Import {
+        ctx: Context,
+        mutations: Vec<Mutation>,
```
Comment (disksing) [inline src/storage/mod.rs]: Maybe `Vec<(Key, Value)>` is more simple.

**Comment by AndreMouche** [comment]
## performance result
### Environment

* 1 tikv on one machine
* 1 tidb, 1 pd on one machine
* load data with tool [loader](https://github.com/pingcap/tidb-tools/tree/shirly/tidb_1pc_import/loader),each branch run 35 minutes.

* table schema:
```
desc to_stat_info;
+-------------------------------+--------------+------+------+---------+-------+
| Field                         | Type         | Null | Key  | Default | Extra |
+-------------------------------+--------------+------+------+---------+-------+
| rec_id                        | bigint(20)   | NO   | PRI  | NULL    |       |
| address                       | longtext     | YES  |      | NULL    |       |
| archive_time                  | datetime     | YES  |      | NULL    |       |
| biz_id                        | int(11)      | YES  |      | NULL    |       |
| cancel_time                   | datetime     | YES  |      | NULL    |       |
| cell_id                       | int(11)      | YES  |      | NULL    |       |
| cell_name                     | varchar(40)  | YES  |      | NULL    |       |
| community_id                  | int(11)      | YES  |      | NULL    |       |
| community_name                | varchar(40)  | YES  |      | NULL    |       |
| coordinate_x                  | double       | YES  |      | NULL    |       |
| coordinate_y                  | double       | YES  |      | NULL    |       |
| create_time                   | datetime     | YES  | MUL  | NULL    |       |
| district_id                   | int(11)      | YES  | MUL  | NULL    |       |
| district_name                 | varchar(40)  | YES  |      | NULL    |       |
| duty_grid_code                | varchar(40)  | YES  |      | NULL    |       |
| duty_grid_id                  | int(11)      | YES  |      | NULL    |       |
| duty_grid_name                | varchar(40)  | YES  |      | NULL    |       |
| event_desc                    | longtext     | YES  |      | NULL    |       |
| event_src_id                  | int(11)      | YES  |      | NULL    |       |
| event_src_name                | varchar(40)  | YES  |      | NULL    |       |
| event_type_code               | varchar(40)  | YES  |      | NULL    |       |
| event_type_id                 | int(11)      | YES  | MUL  | NULL    |       |
| event_type_name               | varchar(20)  | YES  |      | NULL    |       |
| main_type_id                  | int(11)      | YES  |      | NULL    |       |
| main_type_name                | varchar(20)  | YES  |      | NULL    |       |
| rec_type_id                   | int(11)      | YES  |      | NULL    |       |
| rec_type_name                 | varchar(40)  | YES  |      | NULL    |       |
| street_id                     | int(11)      | YES  |      | NULL    |       |
| street_name                   | varchar(40)  | YES  |      | NULL    |       |
| sub_type_id                   | int(11)      | YES  |      | NULL    |       |
| sub_type_name                 | varchar(20)  | YES  | MUL  | NULL    |       |
| task_num                      | varchar(40)  | YES  | MUL  | NULL    |       |
| act_property_id               | int(11)      | YES  |      | NULL    |       |
| archive_human_id              | int(11)      | YES  |      | NULL    |       |
| archive_human_name            | varchar(40)  | YES  |      | NULL    |       |
| archive_role_id               | int(11)      | YES  |      | NULL    |       |
| archive_num                   | int(1)       | YES  |      | 0       |       |
| cancel_num                    | int(11)      | YES  |      | NULL    |       |
| cancel_opinion                | longtext     | YES  |      | NULL    |       |
| check_num                     | int(1)       | YES  |      | 0       |       |
| check_patrol_id               | int(11)      | YES  |      | NULL    |       |
| check_patrol_name             | varchar(40)  | YES  |      | NULL    |       |
| check_time                    | datetime     | YES  |      | NULL    |       |
| check_trans_human_id          | int(11)      | YES  |      | NULL    |       |
| check_trans_human_name        | varchar(40)  | YES  |      | NULL    |       |
| check_trans_role_id           | int(11)      | YES  |      | NULL    |       |
| check_trans_num               | int(1)       | YES  |      | 0       |       |
| check_trans_time              | datetime     | YES  |      | NULL    |       |
| classic_flag                  | int(11)      | YES  |      | NULL    |       |
| dispatch_human_id             | int(11)      | YES  |      | NULL    |       |
| dispatch_human_name           | varchar(40)  | YES  |      | NULL    |       |
| dispatch_role_id              | int(11)      | YES  |      | NULL    |       |
| dispatch_num                  | int(1)       | YES  |      | 0       |       |
| dispatch_time                 | datetime     | YES  |      | NULL    |       |
| display_style_id              | int(11)      | YES  |      | NULL    |       |
| dispose_begin_time            | datetime     | YES  |      | NULL    |       |
| dispose_deadline              | datetime     | YES  |      | NULL    |       |
| dispose_end_time              | datetime     | YES  |      | NULL    |       |
| dispose_num                   | int(1)       | YES  |      | 0       |       |
| dispose_opinion               | longtext     | YES  |      | NULL    |       |
| dispose_region_id             | int(11)      | YES  |      | NULL    |       |
| dispose_region_name           | varchar(40)  | YES  |      | NULL    |       |
| dispose_unit_id               | int(11)      | YES  |      | NULL    |       |
| dispose_unit_name             | varchar(40)  | YES  |      | NULL    |       |
| event_grade_id                | int(11)      | YES  |      | NULL    |       |
| event_grade_name              | varchar(20)  | YES  |      | NULL    |       |
| event_level_id                | int(11)      | YES  |      | NULL    |       |
| event_level_name              | varchar(20)  | YES  |      | NULL    |       |
| event_state_id                | int(11)      | YES  |      | NULL    |       |
| event_state_name              | int(11)      | YES  |      | NULL    |       |
| first_unit_id                 | int(11)      | YES  |      | NULL    |       |
| first_unit_name               | varchar(40)  | YES  |      | NULL    |       |
| hang_num                      | int(1)       | YES  |      | 0       |       |
| his_hang_num                  | int(1)       | YES  |      | 0       |       |
| human_archive_num             | int(1)       | YES  |      | 0       |       |
| human_archive_time            | datetime     | YES  |      | NULL    |       |
| intime_human_archive_num      | int(1)       | YES  |      | 0       |       |
| need_human_archive_num        | int(1)       | YES  |      | 0       |       |
| inst_human_id                 | int(11)      | YES  |      | NULL    |       |
| inst_human_name               | varchar(40)  | YES  |      | NULL    |       |
| inst_num                      | int(1)       | YES  |      | 0       |       |
| inst_role_id                  | int(11)      | YES  |      | NULL    |       |
| inst_time                     | datetime     | YES  |      | NULL    |       |
| intime_archive_num            | int(1)       | YES  |      | 0       |       |
| intime_check_num              | int(1)       | YES  |      | 0       |       |
| intime_dispatch_num           | int(1)       | YES  |      | 0       |       |
| intime_dispose_num            | int(1)       | YES  |      | 0       |       |
| intime_inst_num               | int(1)       | YES  |      | 0       |       |
| intime_operate_num            | int(1)       | YES  |      | 0       |       |
| intime_second_dispatch_num    | int(1)       | YES  |      | 0       |       |
| intime_send_check_num         | int(1)       | YES  |      | 0       |       |
| intime_send_verify_num        | int(11)      | YES  |      | NULL    |       |
| intime_supervise_num          | int(1)       | YES  |      | 0       |       |
| intime_to_archive_num         | int(1)       | YES  |      | 0       |       |
| intime_to_dispose_num         | int(1)       | YES  |      | 0       |       |
| intime_verify_num             | int(11)      | YES  |      | NULL    |       |
| need_archive_num              | int(1)       | YES  |      | 0       |       |
| need_check_num                | int(1)       | YES  |      | 0       |       |
| need_dispatch_num             | int(1)       | YES  |      | 0       |       |
| need_dispose_num              | int(1)       | YES  |      | 0       |       |
| need_second_dispatch_num      | int(1)       | YES  |      | 0       |       |
| need_send_check_num           | int(1)       | YES  |      | 0       |       |
| need_send_verify_num          | int(11)      | YES  |      | NULL    |       |
| need_supervise_num            | int(1)       | YES  |      | 0       |       |
| need_verify_num               | int(11)      | YES  |      | NULL    |       |
| not_inst_num                  | int(11)      | YES  |      | NULL    |       |
| not_operate_num               | int(1)       | YES  |      | 0       |       |
| operate_human_id              | int(11)      | YES  |      | NULL    |       |
| operate_human_name            | varchar(40)  | YES  |      | NULL    |       |
| operate_num                   | int(1)       | YES  |      | 0       |       |
| operate_role_id               | int(11)      | YES  |      | NULL    |       |
| operate_time                  | datetime     | YES  |      | NULL    |       |
| overtime_archive_num          | int(1)       | YES  |      | 0       |       |
| overtime_dispose_num          | int(1)       | YES  |      | 0       |       |
| overtime_to_archive_num       | int(1)       | YES  |      | 0       |       |
| overtime_to_dispose_num       | int(1)       | YES  |      | 0       |       |
| patrol_report_num             | int(1)       | YES  |      | 0       |       |
| postpone_num                  | int(1)       | YES  |      | 0       |       |
| public_report_num             | int(1)       | YES  |      | 0       |       |
| rec_disp_num                  | int(11)      | YES  |      | NULL    |       |
| report_num                    | int(1)       | YES  |      | 1       |       |
| report_patrol_id              | int(11)      | YES  |      | NULL    |       |
| report_patrol_name            | varchar(40)  | YES  |      | NULL    |       |
| rework_num                    | int(1)       | YES  |      | 0       |       |
| multi_rework_num              | int(2)       | YES  |      | 0       |       |
| second_dispatch_human_id      | int(11)      | YES  |      | NULL    |       |
| second_dispatch_human_name    | varchar(40)  | YES  |      | NULL    |       |
| second_dispatch_role_id       | int(11)      | YES  |      | NULL    |       |
| second_dispatch_num           | int(1)       | YES  |      | 0       |       |
| second_dispatch_time          | datetime     | YES  |      | NULL    |       |
| second_unit_id                | int(11)      | YES  |      | NULL    |       |
| second_unit_name              | varchar(40)  | YES  |      | NULL    |       |
| send_check_human_id           | int(11)      | YES  |      | NULL    |       |
| send_check_human_name         | varchar(40)  | YES  |      | NULL    |       |
| send_check_human_role_id      | int(11)      | YES  |      | NULL    |       |
| send_check_num                | int(1)       | YES  |      | 0       |       |
| send_check_time               | datetime     | YES  |      | NULL    |       |
| send_verify_human_id          | int(11)      | YES  |      | NULL    |       |
| send_verify_human_name        | varchar(40)  | YES  |      | NULL    |       |
| send_verify_human_role_id     | int(11)      | YES  |      | NULL    |       |
| send_verify_num               | int(11)      | YES  |      | NULL    |       |
| send_verify_time              | datetime     | YES  |      | NULL    |       |
| supervise_human_id            | int(11)      | YES  |      | NULL    |       |
| supervise_human_name          | varchar(40)  | YES  |      | NULL    |       |
| supervise_role_id             | int(11)      | YES  |      | NULL    |       |
| supervise_num                 | int(1)       | YES  |      | 0       |       |
| supervise_time                | datetime     | YES  |      | NULL    |       |
| third_unit_id                 | int(11)      | YES  |      | NULL    |       |
| third_unit_name               | varchar(40)  | YES  |      | NULL    |       |
| to_archive_num                | int(1)       | YES  |      | 0       |       |
| to_dispatch_num               | int(1)       | YES  |      | 0       |       |
| to_dispose_num                | int(1)       | YES  |      | 0       |       |
| to_inst_num                   | int(1)       | YES  |      | 0       |       |
| to_operate_num                | int(1)       | YES  |      | 0       |       |
| to_supervise_num              | int(11)      | YES  |      | NULL    |       |
| valid_patrol_report_num       | int(1)       | YES  |      | 0       |       |
| valid_public_report_num       | int(1)       | YES  |      | 0       |       |
| valid_report_num              | int(1)       | YES  |      | 0       |       |
| verify_num                    | int(11)      | YES  |      | NULL    |       |
| verify_patrol_id              | int(11)      | YES  |      | NULL    |       |
| verify_patrol_name            | varchar(40)  | YES  |      | NULL    |       |
| verify_time                   | datetime     | YES  |      | NULL    |       |
| period_type                   | int(11)      | YES  |      | NULL    |       |
| shuffle_num                   | int(11)      | YES  |      | NULL    |       |
| overtime_dispatch_times       | float        | YES  |      | NULL    |       |
| district_rework_num           | int(1)       | YES  |      | 0       |       |
| multi_district_rework_num     | int(2)       | YES  |      | 0       |       |
| bundle_dispose_deadline       | datetime     | YES  |      | NULL    |       |
| bundle_dispose_overtime_times | float        | YES  |      | 0       |       |
| dispose_overtime_times        | float        | YES  |      | 0       |       |
| dispose_limit                 | float        | YES  |      | NULL    |       |
| dispose_limit_str             | varchar(20)  | YES  |      | NULL    |       |
| stat_flag                     | int(11)      | YES  |      | NULL    |       |
| dispose_used                  | float        | YES  |      | 0       |       |
| bundle_dispose_used           | float        | YES  |      | 0       |       |
| bundle_dispose_limit          | float        | YES  |      | NULL    |       |
| dispose_act_def_id            | int(11)      | YES  |      | NULL    |       |
| spec_type_id                  | int(11)      | YES  |      | NULL    |       |
| hz_road_type_id               | int(11)      | YES  |      | NULL    |       |
| stat_append_flag              | int(11)      | YES  |      | NULL    |       |
| road_id                       | int(11)      | YES  |      | NULL    |       |
| road_name                     | varchar(100) | YES  |      | NULL    |       |
| road_type_id                  | int(11)      | YES  |      | NULL    |       |
| center_town_flag              | int(11)      | YES  |      | NULL    |       |
| overtime_back_times           | int(11)      | YES  |      | 0       |       |
| execute_time                  | datetime     | YES  |      | NULL    |       |
| new_inst_cond_id              | int(11)      | YES  |      | NULL    |       |
| new_inst_cond_name            | varchar(80)  | YES  |      | NULL    |       |
| archive_cond_id               | int(11)      | YES  |      | NULL    |       |
| archive_cond_name             | varchar(80)  | YES  |      | NULL    |       |
| part_code                     | varchar(40)  | YES  |      | NULL    |       |
| biz_name                      | varchar(40)  | YES  |      | NULL    |       |
| multi_overtime_back_num       | int(2)       | YES  |      | 0       |       |
| process_district_id           | int(11)      | YES  |      | NULL    |       |
| process_district_name         | varchar(40)  | YES  |      | NULL    |       |
| not_operate_opinion           | longtext     | YES  |      | NULL    |       |
| not_operate_type              | int(11)      | YES  |      | 0       |       |
| error_check_flag              | int(11)      | YES  |      | 0       |       |
| event_marks                   | varchar(200) | YES  |      | NULL    |       |
| archive_type_id               | int(11)      | YES  |      | NULL    |       |
| archive_type_name             | varchar(20)  | YES  |      | NULL    |       |
+-------------------------------+--------------+------+------+---------+-------+
```

### Result

The master  branch insert 744840 records
The current branch insert 1201911 records

**Comment by disksing** [comment]
LGTM.

**Comment by BusyJay** [comment]
Can this pr be closed?

**Pull request closed without merging**